src/App.tsx

```diff
@@ -1,24 +1,202 @@
+      })
+      .catch((error) => {
+        setErrorMessage(ErrorMessage.Add);
+        throw error;
+      })
+      .finally(() => {
+        setIsRequesting(false);
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Delete);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleCompletedChange = (todo: Todo) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch((error) => {
+        setErrorMessage(ErrorMessage.Update);
+        throw error;
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleDeleteAllCompletedTodos = () => {
+    todos
+      .filter((todo) => todo.completed)
+      .forEach((todo) => handleDeleteTodo(todo.id));
+  };
+
+  const handleUpdateTodo = (todo: Todo, newTodoTitle: string) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    return todoService
+      .updateTodo({
+        ...todo,
+        title: newTodoTitle,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Update);
+        throw new Error();
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const todosCounter = useMemo(() => {
+    return todos.filter(todo => !todo.completed).length;
+  }, [todos]);
+
+  const isAllCompleted = useMemo(() => {
+    return todos.every(todo => todo.completed);
+  }, [todos]);
+
+  const hasCompletedTodo = todos.some(({ completed }) => completed);
+
+  const handleToggleButton = () => {
+    if (!isAllCompleted) {
+      todos
+        .filter((todo) => !todo.completed)
+        .forEach((todo) => handleCompletedChange(todo));
+    }
+
+    if (isAllCompleted) {
+      todos
+        .filter((todo) => todo.completed)
+        .forEach((todo) => handleCompletedChange(todo));
+    }
```

Comment:
```suggestion
    todos
      .filter((todo) => todo.completed === isAllCompleted)
      .forEach((todo) => handleCompletedChange(todo));
```
You can simplify in this case